linkerd/proxy/http/src/detect.rs

```diff
@@ -0,0 +1,227 @@
+                trace!("Checking H2 preface");
+                if &buf[..H2_PREFACE.len()] == H2_PREFACE {
+                    trace!("Matched HTTP/2 prefix");
+                    return Ok(Some(Version::H2));
+                }
+
+                // Not a match. Don't check for an HTTP/2 preface again.
+                maybe_h2 = false;
+            }
+
+            if maybe_h1 {
+                // Scan up to the first line ending to determine whether the
+                // request is HTTP/1.1.
+                for i in scan_idx..(buf.len() - 1) {
+                    if &buf[i..=i + 1] == b"\r\n" {
+                        trace!(offset = i, "Found newline");
+                        // If the first line looks like an HTTP/2 first line,
+                        // then we almost definitely got a fragmented first
+                        // read. Only try HTTP/1 parsing if it doesn't look like
+                        // HTTP/2.
+                        if !is_h2_first_line(&buf[..=i + 1]) {
+                            trace!("Parsing HTTP/1 message");
+                            // If we get to reading headers (and fail), the
+                            // first line looked like an HTTP/1 request; so
+                            // handle the stream as HTTP/1.
+                            if let Ok(_) | Err(httparse::Error::TooManyHeaders) =
+                                httparse::Request::new(&mut [httparse::EMPTY_HEADER; 0])
+                                    .parse(&buf[..])
+                            {
+                                trace!("Matched HTTP/1");
+                                return Ok(Some(Version::Http1));
+                            }
+                        }
+
+                        // We found the EOL and it wasn't an HTTP/1.x request;
+                        // stop scanning and don't scan again.
+                        maybe_h1 = false;
+                        break;
+                    }
+                }
+
+                // Advance our scan index to the end of buffer so the next
+                // iteration starts scanning where we left off.
+                scan_idx = buf.len() - 1;
+            }
+
+            if !maybe_h1 && !maybe_h2 {
+                trace!("Not HTTP");
+                return Ok(None);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bytes::BufMut;
+    use tokio_test::io;
+
+    const HTTP11_LINE: &'static [u8] = b"GET / HTTP/1.1\r\n";
+    const TIMEOUT: time::Duration = time::Duration::from_secs(10);
+    const H2_AND_GARBAGE: &'static [u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\ngarbage";
+    const GARBAGE: &'static [u8] =
+        b"garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage";
+
+    #[tokio::test]
+    async fn h2() {
+        let _ = tracing_subscriber::fmt::try_init();
```

Comment:
nit/tioli: if these tests all used a single-threaded tokio executor, we could just give each one its own single-threaded default subscriber:
```suggestion
    use tracing_subscriber::prelude::*;
      
    #[tokio::test(flavor = "current_thread")]
    async fn h2() {
        let _subscriber = tracing_subscriber::fmt().set_default();
```
this isn't hugely different but it does keep the tests completely isolated...

Reply:
also, we may want to use 
```rust
tracing_subscriber::fmt().with_test_writer()
```
here so that the traces get captured by libtest and only printed on failures?